pkg/sbom/spdx/marshal.go

```diff
@@ -389,11 +400,97 @@ func (m *Marshaler) spdxAnnotations(c *core.Component, timeNow string) []spdx.An
 	return annotations
 }
 
-func (m *Marshaler) spdxLicense(c *core.Component) string {
+func (m *Marshaler) spdxLicense(c *core.Component) (string, []*spdx.OtherLicense) {
+	// Only library components contain licenses
+	if c.Type != core.TypeLibrary {
+		return "", nil
+	}
 	if len(c.Licenses) == 0 {
-		return noAssertionField
+		return noAssertionField, nil
+	}
+	return m.normalizeLicenses(c.Licenses)
+}
+
+func (m *Marshaler) normalizeLicenses(licenses []string) (string, []*spdx.OtherLicense) {
+	var otherLicenses = make(map[string]*spdx.OtherLicense) // licenseID -> OtherLicense
+
+	// Save text licenses as OtherLicense
+	for i, license := range licenses {
+		if strings.HasPrefix(license, licensing.LicenseTextPrefix) {
+			otherLicense := m.newOtherLicense(strings.TrimPrefix(license, licensing.LicenseTextPrefix), true)
+			otherLicenses[otherLicense.LicenseIdentifier] = otherLicense
+			licenses[i] = otherLicense.LicenseIdentifier
+		}
+	}
+
+	license := strings.Join(lo.Map(licenses, func(license string, index int) string {
+		// e.g. GPL-3.0-with-autoconf-exception
+		license = strings.ReplaceAll(license, "-with-", " WITH ")
+		license = strings.ReplaceAll(license, "-WITH-", " WITH ")
+		return fmt.Sprintf("(%s)", license)
+	}), " AND ")
+
+	normalizedLicense, err := expression.Normalize(license, licensing.NormalizeLicense, expression.NormalizeForSPDX)
+	if err != nil {
+		// Not fail on the invalid license
+		m.logger.Warn("Unable to marshal SPDX licenses", log.String("license", license))
+		return "", nil
+	}
+
+	license, others := m.processNonSpdxLicenses(normalizedLicense)
+	allOtherLicenses := lo.Assign(otherLicenses, others)
+
+	return license, lo.Ternary(len(allOtherLicenses) > 0, lo.Values(allOtherLicenses), nil)
+}
+
+// newOtherLicense create new OtherLicense for license not included in the SPDX license list
+func (m *Marshaler) newOtherLicense(license string, text bool) *spdx.OtherLicense {
+	otherLicense := spdx.OtherLicense{
+		// ExtractedText and LicenseName are mandatory fields
+		ExtractedText: noAssertionField,
+		LicenseName:   noAssertionField,
+	}
+	if text {
+		otherLicense.ExtractedText = license
+	} else {
+		otherLicense.LicenseName = license
+	}
+	licenseID, err := calcSPDXID(m.hasher, otherLicense)
+	if err != nil {
+		m.logger.Warn("Unable to calculate SPDX licenses ID", log.String("license", license), log.Err(err))
+		return nil
 	}
-	return NormalizeLicense(c.Licenses)
+	otherLicense.LicenseIdentifier = LicenseRefPrefix + "-" + licenseID
```

Comment:
Minor suggestion - you could simplify this by including the "-" in the LicenseRefPrefix definition and just use `LicenseRefPrefix + licenseID`  The string "LicenseRef" will likely never be used without the trailing "-".

Reply:
This is our internal decision - we don't use `-` suffixes in constants :)